File: qasdk/src/main/java/co/astrnt/qasdk/core/InterviewObserver.java
package co.astrnt.qasdk.core;

import co.astrnt.qasdk.dao.InformationApiDao;
import co.astrnt.qasdk.dao.InterviewApiDao;
import co.astrnt.qasdk.dao.InterviewResultApiDao;
import co.astrnt.qasdk.dao.LogDao;
import co.astrnt.qasdk.utils.LogUtil;

import static co.astrnt.qasdk.type.InterviewType.ASTRONAUT_PROFILE;
import static co.astrnt.qasdk.type.InterviewType.CLOSE_APTITUDE;
import static co.astrnt.qasdk.type.InterviewType.CLOSE_INTERVIEW;
import static co.astrnt.qasdk.type.InterviewType.CLOSE_INTERVIEW_PROFILE;
import static co.astrnt.qasdk.type.InterviewType.CLOSE_SECTION;
import static co.astrnt.qasdk.type.InterviewType.CLOSE_TEST;
import static co.astrnt.qasdk.type.InterviewType.OPEN;

/**
 * Created by deni rohimat on 09/04/18.
 */
public abstract class InterviewObserver extends MyObserver<InterviewResultApiDao> {

    @Override
    public void onApiResultOk(InterviewResultApiDao resultApiDao) {
        InterviewApiDao data = resultApiDao.getInterview();
        if (data == null) {
            onApiResultError("", "Code not found or interview already finished", "error");
        } else {

            astrntSDK.saveInterviewResult(resultApiDao, data, false);
            if (data.getType().contains(OPEN)) {

                if (data.getInterviewCode() != null) {
                    LogUtil.addNewLog(data.getInterviewCode(),
                            new LogDao("Response API",
                                    "Success, will move to Register"
                            )
                    );
                }
                onNeedToRegister(data);
            } else {
                astrntSDK.saveInterview(data, resultApiDao.getToken(), data.getInterviewCode());
                InformationApiDao information = resultApiDao.getInformation();

                if (information != null && information.isFinished()) {
                    onApiResultError("", information.getMessage(), "error");
                } else {
                    String interviewCode = astrntSDK.getInterviewCode();
                    switch (data.getType()) {
                        case CLOSE_INTERVIEW:
                        case CLOSE_INTERVIEW_PROFILE:
                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Response API",
                                                "Success, will move to Video Interview"
                                        )
                                );
                            }
                            onInterviewType(data);
                            break;
                        case CLOSE_TEST:
                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Response API",
                                                "Success, will move to MCQ Interview"
                                        )
                                );
                            }
                            onTestType(data);
                            break;
                        case CLOSE_SECTION:
                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Response API",
                                                "Success, will move to Section Interview"
                                        )
                                );
                            }
                            onSectionType(data);
                            break;
                        case ASTRONAUT_PROFILE:
                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Response API",
                                                "Success, will move to Astronaut Profile"
                                        )
                                );
                            }
                            onAstronautProfileType(data);
                            break;
                        case CLOSE_APTITUDE:
                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Response API",
                                                "Success, will move to Close Aptitude Rating Scale"
                                        )
                                );
                            }
                            onAptitudeType(data);
                            break;
                        default:
                            if (resultApiDao.getTitle() != null) {
                                onApiResultError(resultApiDao.getTitle(), resultApiDao.getMessage(), "error");
                            } else {
                                onApiResultError("", resultApiDao.getMessage(), "error");
                            }

                            if (interviewCode != null) {
                                LogUtil.addNewLog(interviewCode,
                                        new LogDao("Enter Code Response API",
                                                "Error : " + resultApiDao.getMessage()
                                        )
                                );
                            }
                            break;
                    }
                }
            }
        }
    }

    public abstract void onNeedToRegister(InterviewApiDao interview);

    public abstract void onInterviewType(InterviewApiDao interview);

    public abstract void onTestType(InterviewApiDao interview);

    public abstract void onSectionType(InterviewApiDao interview);

    public abstract void onAstronautProfileType(InterviewApiDao interview);

    public abstract void onAptitudeType(InterviewApiDao interview);

}

File: qasdk/src/main/java/co/astrnt/qasdk/core/InterviewObserver.kt
package co.astrnt.qasdk.core

import co.astrnt.qasdk.dao.InterviewApiDao
import co.astrnt.qasdk.dao.InterviewResultApiDao
import co.astrnt.qasdk.dao.LogDao
import co.astrnt.qasdk.type.InterviewType.ASTRONAUT_PROFILE
import co.astrnt.qasdk.type.InterviewType.CLOSE_APTITUDE
import co.astrnt.qasdk.type.InterviewType.CLOSE_INTERVIEW
import co.astrnt.qasdk.type.InterviewType.CLOSE_INTERVIEW_PROFILE
import co.astrnt.qasdk.type.InterviewType.CLOSE_SECTION
import co.astrnt.qasdk.type.InterviewType.CLOSE_TEST
import co.astrnt.qasdk.type.InterviewType.OPEN
import co.astrnt.qasdk.utils.LogUtil.addNewLog

abstract class InterviewObserver : MyObserver<InterviewResultApiDao>() {

    override fun onApiResultOk(resultApiDao: InterviewResultApiDao) {
        val data = resultApiDao.interview
        if (data == null) {
            onApiResultError("", "Code not found or interview already finished", "error")
        } else {
            astrntSDK.saveInterviewResult(resultApiDao, data, false)
            if (data.type!!.contains(OPEN)) {
                if (data.interviewCode != null) {
                    addNewLog(data.interviewCode,
                            LogDao("Response API",
                                    "Success, will move to Register"
                            )
                    )
                }
                onNeedToRegister(data)
            } else {
                astrntSDK.saveInterview(data, resultApiDao.token, data.interviewCode)
                val information = resultApiDao.information
                if (information != null && information.isFinished) {
                    onApiResultError("", information.message, "error")
                } else {
                    val interviewCode = astrntSDK.interviewCode
                    when (data.type) {
                        CLOSE_INTERVIEW, CLOSE_INTERVIEW_PROFILE -> {
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Response API",
                                                "Success, will move to Video Interview"
                                        )
                                )
                            }
                            onInterviewType(data)
                        }
                        CLOSE_TEST -> {
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Response API",
                                                "Success, will move to MCQ Interview"
                                        )
                                )
                            }
                            onTestType(data)
                        }
                        CLOSE_SECTION -> {
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Response API",
                                                "Success, will move to Section Interview"
                                        )
                                )
                            }
                            onSectionType(data)
                        }
                        ASTRONAUT_PROFILE -> {
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Response API",
                                                "Success, will move to Astronaut Profile"
                                        )
                                )
                            }
                            onAstronautProfileType(data)
                        }
                        CLOSE_APTITUDE -> {
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Response API",
                                                "Success, will move to Close Aptitude Rating Scale"
                                        )
                                )
                            }
                            onAptitudeType(data)
                        }
                        else -> {
                            if (resultApiDao.title != null) {
                                onApiResultError(resultApiDao.title.toString(), resultApiDao.message.toString(), "error")
                            } else {
                                onApiResultError("", resultApiDao.message.toString(), "error")
                            }
                            if (interviewCode != null) {
                                addNewLog(interviewCode,
                                        LogDao("Enter Code Response API",
                                                "Error : " + resultApiDao.message
                                        )
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    abstract fun onNeedToRegister(interview: InterviewApiDao)
    abstract fun onInterviewType(interview: InterviewApiDao)
    abstract fun onTestType(interview: InterviewApiDao)
    abstract fun onSectionType(interview: InterviewApiDao)
    abstract fun onAstronautProfileType(interview: InterviewApiDao)
    abstract fun onAptitudeType(interview: InterviewApiDao)
}